proposals/0185-vote-account-v4.md

```diff
@@ -0,0 +1,243 @@
+    pub block_tips_commission: u8,
+
+    /// NEW: bump seed for deriving this vote accounts stake rewards pool address
+    pub stake_rewards_pool_bump_seed: u8,
+
+    /// REMOVED
+    /// prior_voters: CircBuf<(Pubkey, Epoch, Epoch)>,
+
+    pub votes: VecDeque<LandedVote>,
+    pub root_slot: Option<Slot>,
+    pub authorized_voters: AuthorizedVoters,
+    pub epoch_credits: Vec<(Epoch, u64, u64)>,
+    pub last_timestamp: BlockTimestamp,
+}
+```
+
+Whenever a vote account is modified by the vote program in a transaction AND
+hasn't been updated to v4 yet, the account state MUST be saved in the new format
+with the following default values for the new fields described above:
+
+```rust
+VoteStateV4 {
+    // ..
+
+    inflation_rewards_collector: vote_state_v3.node_pubkey,
+    block_fees_collector: vote_state_v3.node_pubkey,
+    block_tips_collector: vote_state_v3.node_pubkey,
+    inflation_rewards_commission: vote_state_v3.commission,
+    block_fees_commission: 100u8,
+    block_tips_commission: 100u8,
+    stake_rewards_pool_bump_seed: find_stake_rewards_pool_bump_seed(vote_pubkey),
+
+    // ..
+}
+
+fn find_stake_rewards_pool_bump_seed(vote_pubkey: &Pubkey) -> u8 {
+    Pubkey::find_program_address(
+        [
+            b"stake_rewards_pool",
+            vote_pubkey.as_ref(),
+        ],
+        &stake_program::id(),
+    ).1
+}
+```
+
+### Vote Program
+
+```rust
+pub enum VoteInstruction {
+    // ..
+    UpdateCommission {..} // 5u32
+    // ..
+    UpdateCommissionWithKind { // 16u32
+        commission: u8,
+        kind: CollectorKind,
+    },
+    UpdateCollectorAccount { // 17u32
+        pubkey: Pubkey,
+        kind: CollectorKind,
+    },
+}
+
+#[repr(u8)]
+pub enum CollectorKind {
+    InflationRewards = 0,
+    BlockFees,
+    BlockTips,
+}
+```
+
+#### `UpdateCommission`
+
+The existing `UpdateCommission` instruction (with enum discriminant `5u32`) will
+continue to exist but will continue to only update the inflation rewards
+commission.
+
+#### `UpdateCommissionWithKind`
+
+A new instruction for setting different kinds of commissions will be added to
+the vote program with the enum discriminant of `16u32` little endian encoded in
+the first 4 bytes.
+
+#### `UpdateCollectorAccount`
+
+A new instruction for setting collector accounts will be added to the vote
+program with the enum discriminant value of `17u32` little endian encoded in the
+first 4 bytes.
+
+#### `Authorize`, `AuthorizeChecked`, `AuthorizeWithSeed`, `AuthorizeCheckedWithSeed`
+
+Existing authorize instructions will be processed differently when setting new
+authorized voters. Rather than purging authorized voter entries from the
+`authorized_voters` field that correspond to epochs less than the current epoch,
+only purge entries less than the previous epoch (current epoch - 1). This will
+mean that the `authorized_voters` field can now hold up to 4 entries for the
+epochs in the range `[current_epoch - 1, current_epoch + 2]`. Keeping the
```

Comment:
is there a reason we need to store `current_epoch + 2`?

Reply:
Yeah that's the current design and I think it kinda makes sense because otherwise you could set a new authorized voter at the end of an epoch and immediately start using it in the next block in the new epoch and leaders would need an up to date view on any new authorized voters for each fork crossing the epoch boundary